local/utils.sh

```diff
@@ -19,9 +19,17 @@ wait_command() {
   local wait_seconds="${1:-40}"; shift # 40 seconds as default timeout
 
   until [[ $((wait_seconds--)) -eq 0 ]] || eval "$command &> /dev/null" ; do sleep 1; done
+
+  ((++wait_seconds))
```

Comment:
I'm not sure if adding this line is helpful.

Reply:
+1

Reply:
Without this line `wait_command` will always succeed even if the command fails after the timeout.
Which means for example the script [continue here](https://github.com/openshift-pipelines/pipelines-service/blob/43cdfd1c247b5538b68ab828a8695a6d13081eb6/local/kcp/start.sh#L61) even if kcp startup failed.
There are different ways to fix this, I have used the one from the [original kcp script](https://github.com/kcp-dev/kcp/blob/ac174c63a28feec6eaaa6ace6d63528b2333ad50/contrib/demos-unmaintained/demo/.startUtils#L9). Where if the counter `wait_seconds` is non-zero after the loop (command succeeded), the return status will be 0 (success). Otherwise the return status will be 1 (error).

Reply:
but it should be ((--wait_seconds)), shouldn't it?
I did a quick test
~~~
set -o errexit
set -o nounset

wait_command() {
  local command="$1"; shift
  local wait_seconds="${1:-40}"; shift # 40 seconds as default timeout

  until [[ $((wait_seconds--)) -eq 0 ]] || eval "$command &> /dev/null" ; do sleep 1; done

  echo $wait_seconds
  ((--wait_seconds))
  ## does not work
  ## ((++wait_seconds))
}

echo "step 1"
wait_command "sleep 1s" 30
echo $?
echo "step 2"
wait_command "sleep 5s" 2
echo $?
echo "finished"
~~~

Reply:
according to this loop:
```
  until [[ $((wait_seconds--)) -eq 0 ]] || eval "$command &> /dev/null" ; do sleep 1; done
```
We exit the loop when:
-  the command succeeds 
- or the timeout has expired and in this case we will always have **wait_seconds==-1**
So we need to `((++wait_seconds))` and send the exist status (0 failure, success otherwise).

If we `((--wait_seconds))` then this will always succeed even if the command is failing. For example you can try
```
wait_command "ls /non-existing-file" 3
echo finished
```

In your example:  the 2 timeout in `wait_command "sleep 5s" 2` is a particular case where it will always fail, you can try:
```
wait_command "ls /tmp" 3
echo finished
```
**NB**: `wait_seconds` is kind of a biased name, cause the wait is not in seconds in case the command takes more than a second to return a result and the `sleep` is a good example. So the param name makes sense only with commands taking less than a second to return a result

Reply:
Thanks for the explanation. I did miss the point. Replacing step 2 with the following gives the expected result with ((++wait_seconds)): `wait_command "/tmp/test5/this_file_does_not_exist" 10`
The construct is a bit fragile if the command hangs like with sleep. Unlikely with `ls` a bit more with kubectl. I guess it is similar to watch in this respect.